{"cells":[{"cell_type":"markdown","id":"f383c726","metadata":{"id":"f383c726"},"source":["**Summary:** The following notebook contains the code to solve the DCJ-median problem using a genetic algorithm metaheuristic approach. The code contains functions to perform DCJ operations on genomes, construct the adjacency graph between two genomes and use it to determine the distance between them, read in the data files and translate to our genome representation, perform crossover and mutations, generate an initial population, find DCJ operations on the sorting path from one genome to another, and find a median genome using a genetic algorithm approach incorporating all the aforementioned functions. Our implementation is inteded to find the median ancestral genome for lions, tigers, and snow leopards. Our read file function takes the tsv files from GenBank and preprocesses them for use in the notebook. Attatched in our submission is the sample simulated genome data included with our reference paper. A second function for reading in that data has been included as well. We performed multiple experiments using our big cat genome data, the reference paper's data, and our own simulated data along with varying values for parameters such as max_generations and per_step. Ultimtely, we found our algorithm to be effective, but can suffer from runtime issues due to complex computations arising from operations such as finding DCJ operations on a sorting path between two genomes and creating the adjacency graphs.\n","\n","**Material and Methods:** In this section, we present how we developed our algorithm using a genetic heuristic and verified its accuracy with small verifiable, manufactured data sets. We also show how to prepare genomic data from NCBI for use in our algorithm and illustrate how to tackle major problems that can arise.\n","Background- The DCJ median problem\n","The DCJ median problem involves finding a common ancestor or median genome that minimizes the sum of evolutionary events: inversions, translocations, fissions, and fusions are required to transform a set of input genomes into the median genome (Yancopoulous et al., 2005; Gao et al., 2013). Our algorithm takes three genomes of modern species and attempts to follow the possible rearrangements of each species genome to recreate the ancestral genome. The median of the three genomes is thought to represent the ancestral species. If the distances between three possible genomes were represented as a triangle, the median genome would be the centroid of the triangle (Figure 1).\n","\n","![dcj_triangle.jpeg](attachment:dcj_triangle.jpeg)\n","\n","Figure 1: This figure is a basic representation of the DCJ median problem. Three genomes are placed to create a triangle. The dark line along the edge of the triangle represents the edit distance between the two genomes it connects. In the center of the triangle, there is a black dot, representing the median of all three genomes. The dashed line moving away from the corners of the genomes represents the shortest distance for the genome to get to the median. The colored paths represent the wandering path that the genetic algorithm takes as the genomes approach the median.\n","\n","In DCJ, genomes are represented as signed permutations, where each gene corresponds to a signed integer indicating its orientation on a chromosome (Sturtevant and Novitski, 1941; Sankoff, 1992). The algorithm then iteratively applies DCJ operations to shuffle the gene order, aiming to converge towards a median genome that minimizes the sum of events across all input genomes (Gao et al., 2005). The optimization process continues until a stable median genome is reached, offering insights into the evolutionary history shared among the input genomes.\n","Consider the genome G with a signed ordering of genes g₁, g₂, ..., gₙ. There are four main types of rearrangements the genome can undergo. A reversal operation between indices i and j (where i ≠ j) produces a new genome with a linear ordering of genes g₁, g₂, ..., gᵢ₋₁, gⱼ, gⱼ₋₁, ..., gᵢ, gⱼ₊₁, ..., gₙ. For genomes with more than one chromosome, additional events come into play. These include translocation, where the end of one chromosome is broken and attached to the end of another; fission, where one chromosome splits into two; and fusion, where two chromosomes combine into one. The “double-cutand-join” (DCJ) model explains each genome rearrangement event as a single DCJ operation and is mathematically simpler than handling each event individually.\n","A method of finding shorter paths of rearrangement steps is to use an adjacency graph. Genes i and j are considered adjacent if i is followed by j or -j is followed by -i. The notion of a breakpoint arises when two genes are adjacent in one genome but not in the other. Telomeres at the ends of chromosomes lack adjacency with any other genes.\n","When the genomes undergo DCJ, they are cut at adjacencies of indicated breakpoints. The order of the cuts is indicated by an adjacency graph that analyzes to find the shortest path for improving its similarity to the other genomes.\n","\n","\n","![dcj_adjacency.jpeg](attachment:dcj_adjacency.jpeg)\n","\n","\n","Figure 3: Agency graph. The graph represents four cycles. The bottom row (in blue) represents the target genome. In each iteration, the section of the genome that undergoes the DCJ operation is indicated, rearranging the genome into the target genome.\n","Background-Genetic Algorithms\n","Genetic algorithms were first developed in the 1960s by John Holland and his collaborators but didn’t start to become popular until the late 1980s, when programmers started utilizing them to evolve and adapt their code (Holland, 1975; Koza JR, 1990). Genetic algorithms (GAs) are optimization techniques inspired by the principles of natural selection. Beginning with an initial population of potential solutions represented as individuals with sets of parameters, GAs iteratively evolve solutions to complex optimization problems. The process involves evaluating the fitness of each individual based on a predefined fitness function, selecting individuals for the next generation based on their fitness, and applying genetic operations such as crossover (recombination) and mutation to generate offspring. Crossover mimics genetic recombination, while mutation introduces random changes, fostering diversity. The next generation is formed by combining the individuals with highest “fitness scores” and their offspring. This process continues until termination criteria are met. GAs are particularly effective for problems with large solution spaces and complex landscapes, offering a heuristic approach for finding approximate solutions over successive generations.\n"," Our approach:\n","The GenoCATalyst algorithm is used to find the ancestral genome that is represented by the median point of three genomes. Our approach combined the idea of the DCJ-median problem with a genetic algorithm to imitate real life genome rearrangement events, and a new generation is selected according to their fitness. GenoCATalyst uses a dynamic programming approach to calculate the fitness score for each genome in each iteration of the algorithm. It utilizes an adjacency graph to reduce the distance between each genome and puts forth the best possible candidates each round to estimate closer and closer to the median ancestor.\n","Data collection and pre-processing\n","Representative species genomes used in our data analysis were obtained from NCBI’s genome library. Species genomes were selected based on the availability of chromosome information and NCBI reference genome validation as reference genomes. With the wide availability of sequencing possible today, many species have been sequenced, but not all assemblies provide details on gene orientation or position on chromosomes. The genomes we selected include the following in the big cat family: lion, tiger, snow leopard (panthera leo, panthera tigris, panthera unica). All three of these are in the Panthera genus, and thus reduce the distance of differences between the potential genomic profiles. Their last common ancestor was ~3.5 million years ago (O’Brian, 2007)\n","I. Developing the algorithm\n","Individuals of the initial population have a deep impact on the performance of a genetic-algorithm based approach. In a similar method of Gao et al (2013), we generated our starting population based on the following observation that given three genomes, the median path is more likely to be on one leaf than the other (Moret et al, 2002). We first look at the distances between the three base genomes, and the larger the distance, the larger of an initial population we generate to cover a larger search space. For each pair of genomes, we iterate over distance/10, 2*(distance/10), … ,6*(distance/10) to define the number of steps. We then generate a ‘per_step’ number of genomes on the sorting path from the first genome in the pair, where ‘per_step’ is a variable passed into the initial population generation function that we can define and fine tune ourselves. Genomes are generated on the sorting path by applying random DCJ operations. This is repeated until all three genome pairs have been looped through and results in our starting population. Generating the starting population in this method helps to improve the likelihood that the algorithm will reach conversion (Gao et al 2013). In testing, we attempted to generate enough diversity by generating 50 genomes per sampled step.\n"," Fine tuning the performance of the algorithm involved utilizing the premade Julia functions over loop creations, as the Julia functions were already optimized to operate in the language. During much of the modification of the adjacency graph. Further optimization of the algorithm involved modifying the number of ‘max generations’ that the algorithm went through before termination. The greater the size of ‘max generations’, the longer the algorithm can run, especially if it never converges. Other variables that could be modified for speed vs accuracy trade off include changing the amount of DCJ operations performed in the crossover and mutate functions. Lowering the amount of operations in those functions increases the speed of the algorithm's main loop, at the cost of having a slower rate of convergence. For testing purposes, we also ran the code a few times using the hamming distance to calculate the median score for the fitness function rather than the DCJ distance. This was just to improve the runtime of the algorithm, as the hamming distance is less computationally intensive, and was used mainly to verify the code was running properly elsewise.\n","II. Output/visualization We output a graph of fitness score over number of iterations (Figure 5). This helps the user to see that the algorithm is progressing towards convergence as more iterations occur.\n","\n","![plot.png](attachment:plot.png)\n","\n","Figure 5: Expected output graph illustrating decreasing fitness score over iteration as the algorithm approaches convergence.\n","III. Testing the algorithm We tested its validity by using our DCJ function on a smaller set of simulated data to ensure that it behaved as expected. To compare performance between algorithms, the same simulated genomes were run in GenoCATalyst and ASMedian algorithms. The genomes were simulated with increasing size (quantity of genes) and increasing number of reversals (percent mutated) in order to evaluate how well the algorithms preformed in terms of time, memory and accuracy.\n","\n","**Results:** Our experiments thereby began by determining how genome size affected different performance metrics of the algorithm, such as time, memory and score. We used the ASMedian algorithm developed by Sankoff and Banette (1998) to compare our algorithm’s performance. GenoCATalyst and ASMedian performance was compared over size of genomes (number of genes) for the median score it produced (Figure 6A. B) and compared to the difference between GenoCATalyst and ASMedian scores (Figure 6C)\n","\n","Figure 6: Comparison of the reported median score in [our algorithm] and ASMedian. Top chart depicts GenoCATalyst middle chart depicts the reported median score for the ASMedian algorithm. Colors indicate genome size tested. We also tested and compared the quantity of memory consumed by GenoCATalyst. The results can be seen in figure 7.\n","\n","![dcj_memory.png](attachment:dcj_memory.png)\n","\n","Figure 7: Comparison of GenoCATalyst memory use (MB) vs ASMedian algorithm. Performance times were also compared between GenoCATalyst as compared to the exact ASMedian solver (Figure 8). The ASM median solver has time performance issues when greater than 40% of the genome is mutated, as seen in figure 8a. Figure 8b illustrates the differences in runtime between GenoCATalyst and ASMedian solver when the genetic differences between ancestral and modern genomes are less than 40%\n","\n","![dcj_time_diff.png](attachment:dcj_time_diff.png)\n","\n","Figure 8: Comparison of the difference of GenoCATalyst vs ASMedian time performance.\n","\n","**Discussion:** In this paper, we discussed an algorithm that attempted to reconstruct the ancestral genome of several different large cat species. We used an algorithmic approach for the DCJ median problem, along with a genetic algorithm as a heuristic approach. Ultimately, GenoCATalyst did not perform as we hoped it would. Given more time and resources, we would have tried to improve the algorithm through parallelization of the code so that it would be able to run on multiple cores for improved performance. Ideally that would enable us to estimate the ancestor of lions, tigers, and snow leopards as we had originally intended.\n","When we were finally set up to run our algorithm on the big cat genomic data, we were suddenly faced with the fact that we did not have enough time to process it. The adjacency graphs for each pair of genomes were taking tens of minutes to complete, and we would have to be creating many per generation iteration that the algorithm ran. We continued to try to assess our algorithm with the data permutations from the Gao et al (2013) paper, as well as from simulations of reversions.\n","Our algorithm did find the DCJ distance of our test sequences accurately. Also, the median score calculated for our algorithm vs the ASMedian algorithm were consistent up to point where the ASMedian algorithm starts becoming less efficient (when analyzing more distant genomes). Given more time, we would be hopeful to get our algorithm fully functional.\n","Reconstruction of the ancestral genome can give insights into comparative genetics, which can allow for comparison of the genetic makeup of ancestral species with their modern descendants. This is important to showing how genetics have changed over time and aid in understanding of evolutionary relationships. Additionally, There are a few applications that an ancestral genome could provide to modern conservation efforts. Knowledge of the ancestral genome can help biologists recognize genetic trends that have been preserved in the organism’s history. Understanding the ancestral genome can be utilized in designing conservation strategies to preserve genetic heritage. Conservation plans can be developed to maintain historically significant genes while planning breeding programs to help conserve and maintain endangered species. Knowledge of the ancestral genome could assist geneticists and conservationists to create a gene therapy plan to replace faulty genes with better versions from the past. Someday, solving for the ancestral genome may be the key to maintaining genetic diversity of life on our planet.  \n","\n","**References:**\n","\n","Gao N, Yang N, Tang J (2013) Ancestral genome inference using a genetic algorithm approach. PLOS ONE 8(5): e62156. https://doi.org/10.1371/journal.pone.0062156 [date accessed: 10/23/2023]\n","Holland, J. H. (1975). Adaptation in natural and artificial systems: An introductory analysis with applications to biology, control, and artificial intelligence. U Michigan Press.\n","Koza JR. Genetic programming: A paradigm for genetically breeding populations of computer programs to solve problems. Stanford, CA: Stanford University, Department of Computer Science; 1990 Jun 1.\n"," Moret, B.M.E., Siepel, A.C., Tang, J., Liu, T. (2002). Inversion Medians Outperform Breakpoint Medians in Phylogeny Reconstruction from Gene-Order Data. In: Guigó, R., Gusfield, D. (eds) Algorithms in Bioinformatics. WABI 2002. Lecture Notes in Computer Science, vol 2452. Springer, Berlin, Heidelberg. https://doi.org/10.1007/3-540-45784-4_40\n","O’Brien, J. S. and Johnson, W. E. (2007). The evolution of cats. Scientific American July 2007: 68-75\n","Rajan, Vaibhav et al. “Heuristics for the inversion median problem.” BMC bioinformatics vol. 11 Suppl 1,Suppl 1 S30. 18 Jan. 2010, doi:10.1186/1471-2105-11-S1-S30\n","Rubert, D.P., Feijão, P., Braga, M.D.V. et al. Approximating the DCJ distance of balanced genomes in linear time. Algorithms Mol Biol 12, 3 (2017). https://doi.org/10.1186/s13015-017-0095-y\n","Sankoff, D. Edit distance for genome comparison based on non-local operations. In Apostolico, A., Crochemore, M., Galil, Z., Manber, U. (Eds.). Combinatorial Pattern Matching. Third Annual Symposium, Lecture Notes in Computer Science. 1992. Springer Verlag 644, pp. 121–135.\n"," Sankoff D, Blanchette M (1998) Multiple genome rearrangement and breakpoint phylogeny. Journal of Computational Biology 5: 555–570.\n","Sturtevant, A H, and E Novitski. The Homologies of the Chromosome Elements in the Genus Drosophila. Genetics vol. 26,5 (1941): 517-41. doi:10.1093/genetics/26.5.517 (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1209144/pdf/517.pdf) [date accessed: 11/19/2023]\n","Xu W, Sankoff D (2008). Decompositions of multiple breakpoint graphs and rapid exact solutions to the median problem. In: 8th International Workshop on Algorithms in Bioinformatics (WABI 2008). pp. 25-37.\n","Yancopoulos S, Attie O, Friedberg R, Efficient Sorting of Genomic Permutations by Translocation, Inversion and Block Interchange, Bioinformatics, Volume 21, Issue 16, August 2005, Pages 3340–3346, https://doi.org/10.1093/bioinformatics/bti535 [date accessed: 10/29/2023]\n"]},{"cell_type":"markdown","id":"4228b45c","metadata":{"id":"4228b45c"},"source":["**DCJ Operation:**\n","\n","The crux of the entire process, the dcj_operation function performs a dcj operation on an input genome, and returns the resulting genome. The function takes in four indices, two for whcih chromosomes, and two for the genes in those respective chromosomes. The function is able to handle interchromosomal and intrachromosomal operations. The function can perform operations on both linear and circular chromosomes, by first checking for circulairty in each. The function aslo checks for telomeres in order to appropriately decide if the operation will be one of the following:\n","\n","-Fission\n","\n","-Inversion\n","\n","-Fusion\n","\n","-Translocation"]},{"cell_type":"code","execution_count":null,"id":"d6bfd540","metadata":{"id":"d6bfd540","outputId":"7d8bb0ad-b2c9-416a-8205-d651dd77989d"},"outputs":[{"data":{"text/plain":["dcj_operation (generic function with 1 method)"]},"execution_count":1,"metadata":{},"output_type":"execute_result"}],"source":["function dcj_operation(genome, index1, chromosome1, index2, chromosome2)\n","    # Ensure indices are within bounds\n","    if chromosome1 > length(genome) || chromosome2 > length(genome) ||\n","       index1 > length(genome[chromosome1]) || index2 > length(genome[chromosome2])\n","        error(\"Index out of bounds\")\n","    end\n","\n","    # Swap indices and chromosomes if necessary for consistency\n","    if index1 > index2\n","        index1, index2 = index2, index1\n","        chromosome1, chromosome2 = chromosome2, chromosome1\n","    end\n","\n","    # Clone the genome array to avoid mutating the original\n","    new_genome = deepcopy(genome)\n","\n","    # Check for circularity\n","    is_circular1 = !in(0, new_genome[chromosome1])\n","    is_circular2 = !in(0, new_genome[chromosome2])\n","\n","    if chromosome1 == chromosome2\n","        # Handling the same chromosome\n","        part1 = new_genome[chromosome1][1:index1-1]\n","        part2 = new_genome[chromosome1][index1:index2]\n","        part3 = new_genome[chromosome1][index2+1:end]\n","\n","        if is_circular1\n","            # Handling circular chromosome\n","            new_genome[chromosome1] = vcat(part1, reverse(part2), part3)\n","        else\n","            # Handling linear chromosome\n","            if new_genome[chromosome1][index1] == 0 || new_genome[chromosome1][index2] == 0\n","                # Fission - correctly place telomeres at the ends\n","                part2_start = index1 + (new_genome[chromosome1][index1] == 0 ? 1 : 0)\n","                part2 = new_genome[chromosome1][part2_start:index2-1]\n","                if !isempty(part2)\n","                    new_genome[chromosome1] = vcat(part1, 0)\n","                    push!(new_genome, vcat(0, reverse(part2), part3))\n","                end\n","            else\n","                # Inversion\n","                new_genome[chromosome1] = vcat(part1, reverse(part2), part3)\n","            end\n","        end\n","    else\n","        # Handling different chromosomes\n","        part1 = new_genome[chromosome1][1:index1-1]\n","        part2 = new_genome[chromosome1][index1:end]\n","        part3 = new_genome[chromosome2][1:index2-1]\n","        part4 = new_genome[chromosome2][index2:end]\n","\n","        if is_circular1 || is_circular2\n","            # Handling circular chromosomes\n","            new_genome[chromosome1] = vcat(part1, part4)\n","            new_genome[chromosome2] = vcat(part3, part2)\n","        else\n","            # Handling linear chromosomes\n","            if new_genome[chromosome1][index1] == 0 && new_genome[chromosome2][index2] == 0\n","                # Fusion - join chromosomes without telomeres in between\n","                if length(part1) > 1 && length(part4) > 1\n","                    new_genome[chromosome1] = vcat(part1, part4[2:end])\n","                    new_genome[chromosome2] = []\n","                elseif length(part2) > 1 && length(part3) > 1\n","                    new_genome[chromosome1] = vcat(part3[1:end-1], part2)\n","                    new_genome[chromosome2] = []\n","                end\n","            else\n","                # Translocation\n","                new_genome[chromosome1] = vcat(part1, part4)\n","                new_genome[chromosome2] = vcat(part3, part2)\n","            end\n","        end\n","    end\n","\n","    # Remove any empty chromosomes\n","    new_genome = filter(chromosome -> !isempty(chromosome), new_genome)\n","\n","    return new_genome\n","end\n"]},{"cell_type":"markdown","id":"6c0a78bc","metadata":{"id":"6c0a78bc"},"source":["**Adjacency Graph:**\n","\n","In order to properly determine the dcj distance between two genomes, the corresponding adjacency graph has to be constucted. This function creates a graph using the LightGraphs package and adds edges between the adjacencies in each genome. This function can be quite computationally expensive when having to handle large inputs, and has proven to be the main bottleneck in running the code."]},{"cell_type":"code","execution_count":null,"id":"38f2fd2d","metadata":{"id":"38f2fd2d","outputId":"3f77c587-b90e-4462-ce0e-fe4911f6d434"},"outputs":[{"data":{"text/plain":["adjacency_graph (generic function with 1 method)"]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["using LightGraphs\n","\n","function adjacency_graph(genome1, genome2)\n","    # Function to create adjacency pairs from a chromosome\n","    function get_adjacency_pairs(chromosome)\n","        adjacencies = []\n","        n = length(chromosome)\n","        is_circular = chromosome[1] != 0 && chromosome[end] != 0\n","\n","        if is_circular\n","            push!(adjacencies, (chromosome[end], -chromosome[1]))\n","        elseif chromosome[1] != 0\n","            push!(adjacencies, (0, chromosome[1]))\n","        end\n","\n","        for i in 1:(n - 1)\n","            push!(adjacencies, (chromosome[i], -chromosome[i + 1]))\n","        end\n","\n","        if !is_circular && chromosome[end] != 0\n","            push!(adjacencies, (chromosome[end], 0))\n","        end\n","\n","        return adjacencies\n","    end\n","\n","    # Collect all adjacencies from both genomes\n","    adjacencies_genome1 = [get_adjacency_pairs(chromosome) for chromosome in genome1]\n","    adjacencies_genome2 = [get_adjacency_pairs(chromosome) for chromosome in genome2]\n","\n","    # Flatten the lists for easier processing\n","    flat_adjacencies_genome1 = vcat(adjacencies_genome1...)\n","    flat_adjacencies_genome2 = vcat(adjacencies_genome2...)\n","\n","    # Create vertices and initialize graph\n","    total_adjacencies = vcat(flat_adjacencies_genome1, flat_adjacencies_genome2)\n","    #sg = SimpleGraph(length(total_adjacencies))\n","    #g = MetaGraph(length(total_adjacencies))\n","    g = SimpleDiGraph(length(total_adjacencies))\n","\n","    # Create a mapping from adjacencies to vertex indices\n","    vertex_map_genome1 = Dict(adjacency => i for (i, adjacency) in enumerate(flat_adjacencies_genome1))\n","    offset = length(flat_adjacencies_genome1)\n","    vertex_map_genome2 = Dict(adjacency => i + offset for (i, adjacency) in enumerate(flat_adjacencies_genome2))\n","\n","    # Add edges based on shared non-telomere elements\n","    for adjacency1 in flat_adjacencies_genome1\n","        index1 = vertex_map_genome1[adjacency1]\n","        for adjacency2 in flat_adjacencies_genome2\n","            index2 = vertex_map_genome2[adjacency2]\n","\n","            if adjacency1[1] != 0 && (adjacency1[1] == adjacency2[1] || adjacency1[1] == adjacency2[2])\n","                add_edge!(g, index1, index2)\n","            end\n","            if adjacency1[2] != 0 && (adjacency1[2] == adjacency2[1] || adjacency1[2] == adjacency2[2])\n","                add_edge!(g, index2, index1)\n","            end\n","        end\n","    end\n","\n","    return g, vertex_map_genome1, vertex_map_genome2\n","end\n"]},{"cell_type":"markdown","id":"d1236a69","metadata":{"id":"d1236a69"},"source":["Example vertex mappings and edges from the adjacency graph following the example genomes from the slides"]},{"cell_type":"code","execution_count":null,"id":"da9dadb8","metadata":{"id":"da9dadb8","outputId":"cb6b5d3e-749c-41ba-ffca-a356a87528d9"},"outputs":[{"name":"stdout","output_type":"stream","text":["Dict((4, 0) => 5, (1, 2) => 2, (5, -6) => 7, (0, -1) => 1, (7, -5) => 6, (6, -7) => 8, (3, -4) => 4, (-2, -3) => 3)\n","Dict((1, -2) => 10, (5, -6) => 15, (7, 0) => 17, (0, -1) => 9, (4, -5) => 14, (2, -3) => 11, (6, -7) => 16, (3, 0) => 12, (0, -4) => 13)\n","Edge 2 => 10\n","Edge 3 => 10\n","Edge 4 => 12\n","Edge 5 => 14\n","Edge 6 => 17\n","Edge 7 => 15\n","Edge 8 => 16\n","Edge 9 => 1\n","Edge 11 => 2\n","Edge 11 => 3\n","Edge 13 => 4\n","Edge 14 => 6\n","Edge 15 => 7\n","Edge 16 => 8\n"]}],"source":["A = [[0, 1, -2, 3, 4, 0],[5, 6, 7]]\n","B = [[0, 1, 2, 3, 0],[0, 4, 5, 6, 7, 0]]\n","\n","graph, vertex_map1, vertex_map2 = adjacency_graph(A, B)\n","\n","println(vertex_map1)\n","println(vertex_map2)\n","for e in edges(graph)\n","    println(e)\n","end\n"]},{"cell_type":"markdown","id":"b7a8f843","metadata":{"id":"b7a8f843"},"source":["**Get Operations:**\n","\n","This function returns genome that is max_operations on the sorting path from the source genome to the target genome. This is used in the crossover and mutate portions of the genetic algorithm following the process described in the reference paper. This function is an adaptation of the pseudocode for the DCJSORT algorithm discussed in class. This function has also proved to be a bottleneck when running the code, so different ranges of values for max_operations can be used to speed up the overall runtime with the tradeoff of slowing down the rate of convergence."]},{"cell_type":"code","execution_count":null,"id":"170ba31c","metadata":{"id":"170ba31c","outputId":"b4399f7f-9e16-42d1-e315-ed903d7461fb"},"outputs":[{"data":{"text/plain":["get_operations (generic function with 1 method)"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["function get_operations(source, target, max_operations)\n","    modified_source = deepcopy(source)\n","    operation_count = 0\n","\n","    while operation_count < max_operations\n","        operation_found = false\n","\n","        # Process non-telomere adjacencies in target genome\n","        for t_chr_idx in 1:length(target)\n","            for t_gene_idx in 1:(length(target[t_chr_idx]) - 1)\n","                p = target[t_chr_idx][t_gene_idx]\n","                q = target[t_chr_idx][t_gene_idx + 1]\n","\n","                if p != 0 && q != 0\n","                    u_chromosome = v_chromosome = -1\n","                    u_index = v_index = -1\n","\n","                    # Find positions of p and q in modified_source genome\n","                    for s_chr_idx in 1:length(modified_source)\n","                        for s_gene_idx in 1:(length(modified_source[s_chr_idx]) - 1)\n","                            if modified_source[s_chr_idx][s_gene_idx] == p\n","                                u_chromosome, u_index = s_chr_idx, s_gene_idx\n","                            end\n","                            if modified_source[s_chr_idx][s_gene_idx + 1] == q\n","                                v_chromosome, v_index = s_chr_idx, s_gene_idx + 1\n","                            end\n","                        end\n","                    end\n","\n","                    if u_chromosome != -1 && v_chromosome != -1\n","                        # Apply the operation\n","                        modified_source = dcj_operation(modified_source, u_index, u_chromosome, v_index, v_chromosome)\n","                        operation_found = true\n","                        operation_count += 1\n","                        break  # Break to recompute the adjacency graph\n","                    end\n","                end\n","            end\n","            if operation_found\n","                break\n","            end\n","        end\n","\n","        if !operation_found\n","            # Process telomeres in the target genome\n","            for t_chr_idx in 1:length(target)\n","                for t_gene_idx in [1, length(target[t_chr_idx])]\n","                    p = target[t_chr_idx][t_gene_idx]\n","                    if p != 0\n","                        u_chromosome = -1\n","                        u_index = -1\n","\n","                        # Find the chromosome in modified_source with the corresponding gene\n","                        for s_chr_idx in 1:length(modified_source)\n","                            for s_gene_idx in 1:length(modified_source[s_chr_idx])\n","                                if modified_source[s_chr_idx][s_gene_idx] == p\n","                                    u_chromosome, u_index = s_chr_idx, s_gene_idx\n","                                    # Apply the operation\n","                                    modified_source = dcj_operation(modified_source, u_index, u_chromosome, u_index, u_chromosome)\n","                                    operation_count += 1\n","                                    operation_found = true\n","                                    break\n","                                end\n","                            end\n","                            if operation_found\n","                                break\n","                            end\n","                        end\n","                    end\n","                    if operation_found\n","                        break\n","                    end\n","                end\n","            end\n","        end\n","\n","        if !operation_found\n","            # No more operations can be found\n","            break\n","        end\n","    end\n","\n","    return modified_source\n","end\n"]},{"cell_type":"markdown","id":"28465d17","metadata":{"id":"28465d17"},"source":["**DCJ Distance:**\n","\n","One crucial aspect is making sure the dcj distance is properly computed so that the fitness function can provide an accurate score. This function follows the dcj distance equation can looks for the number of cycles and odd-edge paths in the adjacency graph between the two genomes."]},{"cell_type":"code","execution_count":null,"id":"9ae1ee17","metadata":{"id":"9ae1ee17","outputId":"9688476e-3537-415a-ede5-3fd3ab07f084"},"outputs":[{"data":{"text/plain":["dcj_distance (generic function with 1 method)"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["using LightGraphs\n","\n","function dcj_distance(genome1, genome2)\n","    g, _ = adjacency_graph(genome1, genome2)\n","    components = connected_components(g)\n","\n","    C = 0  # Number of cycles of length 2\n","    I = 0  # Number of odd edge paths\n","\n","    for component in components\n","        # Check if the component forms a cycle of length 2\n","        if all(v -> degree(g, v) == 2, component)\n","            C += 1\n","        else\n","            I += 1\n","        end\n","    end\n","\n","    n = sum(count(x -> x != 0, chromosome) for chromosome in genome1)\n","    return n - (C + I ÷ 2)\n","end\n","\n","\n"]},{"cell_type":"markdown","id":"ff19acce","metadata":{"id":"ff19acce"},"source":["**Initial Population:**\n","\n","Generates the initial population to be used in the genetic algorithm. The genomes are generated on a random distance from thee genomes we're finding the median for. The size of the initial population varies depending on the distance between the three genomes."]},{"cell_type":"code","execution_count":null,"id":"50492b3e","metadata":{"id":"50492b3e","outputId":"62521e3a-5098-4838-a3e7-29cdda6013bf"},"outputs":[{"data":{"text/plain":["initial_population (generic function with 1 method)"]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":["function initial_population(genomes, per_step)\n","    population = []\n","\n","    # Calculate DCJ distances between all pairs of genomes\n","    distances = [dcj_distance(genomes[1], genomes[2]),\n","                 dcj_distance(genomes[1], genomes[3]),\n","                 dcj_distance(genomes[2], genomes[3])]\n","\n","    # Iterate over all pairs of genomes\n","    j = 1\n","    for i in 1:3\n","        genome_pair = i != 3 ? (genomes[1], genomes[i+1]) : (genomes[2], genomes[3])\n","\n","        # Calculate steps and iterate over them\n","        for step in ceil(Int, distances[i]/10):ceil(Int, distances[i]/10):floor(Int, 6*distances[i]/10)\n","            # Generate per_step genomes for each step\n","            for _ in 1:per_step\n","                new_genome = deepcopy(genome_pair[1])\n","                # Apply DCJ operations step times to generate a new genome\n","                for _ in 1:step\n","                    chromosome_index1 = rand(1:length(new_genome))\n","                    gene_index1 = rand(1:length(new_genome[chromosome_index1]))\n","\n","                    # Ensure the second set of indices is different from the first\n","                    chromosome_index2, gene_index2 = chromosome_index1, gene_index1\n","                    while chromosome_index2 == chromosome_index1 && gene_index2 == gene_index1\n","                        chromosome_index2 = rand(1:length(new_genome))\n","                        gene_index2 = rand(1:length(new_genome[chromosome_index2]))\n","                    end\n","                    new_genome = dcj_operation(new_genome, gene_index1, chromosome_index1, gene_index2, chromosome_index2)\n","                end\n","                println(j)\n","                j = j + 1\n","                push!(population, new_genome)\n","            end\n","        end\n","    end\n","    return population\n","end\n"]},{"cell_type":"markdown","id":"3d2fb05c","metadata":{"id":"3d2fb05c"},"source":["Finds the perfect median score of the three genomes based off the equation given in the reference paper"]},{"cell_type":"code","execution_count":null,"id":"5122cf40","metadata":{"id":"5122cf40","outputId":"2e2ea002-d3cd-4e7c-fb75-0f6bd79ee0a7"},"outputs":[{"data":{"text/plain":["best_score (generic function with 1 method)"]},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["function best_score(genome1, genome2, genome3)\n","    distance12 = dcj_distance(genome1, genome2)\n","    distance13 = dcj_distance(genome1, genome3)\n","    distance23 = dcj_distance(genome2, genome3)\n","\n","    return ceil((distance12 + distance13 + distance23)/2)\n","end"]},{"cell_type":"markdown","id":"365e52f5","metadata":{"id":"365e52f5"},"source":["Hamming distance function used for testing purposes"]},{"cell_type":"code","execution_count":null,"id":"fd2e0cc5","metadata":{"id":"fd2e0cc5","outputId":"9921bad4-e587-4fb3-f4d9-b5a6fd10df87"},"outputs":[{"data":{"text/plain":["hamming_distance (generic function with 1 method)"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["function hamming_distance(genome1, genome2)\n","    max_chromosomes = max(length(genome1), length(genome2))\n","    total_distance = 0\n","\n","    for i in 1:max_chromosomes\n","        chromo1 = i <= length(genome1) ? genome1[i] : []\n","        chromo2 = i <= length(genome2) ? genome2[i] : []\n","        max_length = max(length(chromo1), length(chromo2))\n","\n","        for j in 1:max_length\n","            if j > length(chromo1) || j > length(chromo2) || chromo1[j] != chromo2[j]\n","                total_distance += 1\n","            end\n","        end\n","    end\n","\n","    return total_distance\n","end\n"]},{"cell_type":"markdown","id":"a1d103a5","metadata":{"id":"a1d103a5"},"source":["**Median Score:**\n","\n","Calculates the median score of the genome against the three genomes we're finding the median for. This function is essential for finding the fitness score of each genome in the population. However, due to how complex computing the dcj distance can be, this function can become a major bottleneck, and a large amount of time can be spent on repeated calls to this function on every iteration. As such, a supplemental version is supplied in the commented out portion that computes using the hamming distance rather than the dcj distance. This does not provide an accurate score, but can be used to speed up the overall runtime to verify that the genetic algorithm works, and as such is mainly used for testing purposes."]},{"cell_type":"code","execution_count":null,"id":"f1d05cc7","metadata":{"id":"f1d05cc7","outputId":"b3978c30-e144-4401-bc4a-67ed73174ebe"},"outputs":[{"data":{"text/plain":["median_score (generic function with 1 method)"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["using Statistics\n","\n","function median_score(genome1, genome2, genome3, m)\n","    distance1 = dcj_distance(genome1, m)\n","    distance2 = dcj_distance(genome2, m)\n","    distance3 = dcj_distance(genome3, m)\n","\n","    # Calculate the median of the three distances\n","    return median([distance1, distance2, distance3])\n","end\n","\n","#=\n","using Statistics\n","\n","function median_score(genome1, genome2, genome3, m)\n","    distance1 = hamming_distance(genome1, m)\n","    distance2 = hamming_distance(genome2, m)\n","    distance3 = hamming_distance(genome3, m)\n","\n","    # Calculate the median of the three distances\n","    return median([distance1, distance2, distance3])\n","end\n","=#"]},{"cell_type":"markdown","id":"e61133ef","metadata":{"id":"e61133ef"},"source":["**Fitness:**\n","\n","This function returns the fitness score of the genome by implementing the equation (N-(S-Sbest)) given in the reference paper"]},{"cell_type":"code","execution_count":null,"id":"9284747e","metadata":{"id":"9284747e","outputId":"904928a5-558b-482e-d23d-a18e7540656e"},"outputs":[{"data":{"text/plain":["fitness (generic function with 1 method)"]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["function fitness(genome1, genome2, genome3, m, best)\n","    #N = count(x -> x > 0, m)\n","    N = sum(count(x -> x > 0, chromosome) for chromosome in m)\n","    S = median_score(genome1, genome2, genome3, m)\n","\n","    return N - (S - best)\n","end"]},{"cell_type":"markdown","id":"c99a9e13","metadata":{"id":"c99a9e13"},"source":["**Get Scores:**\n","\n","Returns the entire population sorted by their fitness scores in descending order for processing purposes in the genetic algorithm"]},{"cell_type":"code","execution_count":null,"id":"f9c22f5f","metadata":{"scrolled":true,"id":"f9c22f5f","outputId":"0ba4833c-657f-4b8d-b201-ff75748ae770"},"outputs":[{"data":{"text/plain":["get_scores (generic function with 1 method)"]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["function get_scores(genomes, pop, best)\n","    #fit_dict = Dict{Vector{Int64}, Int}()\n","    fit_dict = Dict{Vector{Vector{Int}}, Int}()\n","\n","    for m in pop\n","       fit_dict[m] = fitness(genomes[1], genomes[2], genomes[3], m, best)\n","        print('-')\n","        flush(stdout)\n","    end\n","\n","    # Sort the dictionary by values in descending order and return the keys\n","    sorted_genomes = [kv[1] for kv in sort(collect(fit_dict), by = x -> x[2], rev = true)]\n","    println()\n","\n","    return fit_dict, sorted_genomes\n","end\n","\n","\n"]},{"cell_type":"markdown","id":"905aac56","metadata":{"id":"905aac56"},"source":["**Crossover:**\n","\n","Crossover creates two child genomes from two parent genomes to be added to the new population. The fitness scores of the two parents are measured in order to determine a better parent and a worse parent. The first child, C1, is a direct copy of the better parent, while the second child, C2, is generated on the sorting path between the worse parent and the better parent."]},{"cell_type":"code","execution_count":null,"id":"3ebaea9e","metadata":{"id":"3ebaea9e","outputId":"e7667367-09f7-4cf0-8cb8-7e358ee2d08c"},"outputs":[{"data":{"text/plain":["crossover (generic function with 1 method)"]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["function crossover(parent1, parent2, genome1, genome2, genome3, best)\n","    # Calculate fitness scores for both parents\n","    F1 = fitness(genome1, genome2, genome3, parent1, best)\n","    F2 = fitness(genome1, genome2, genome3, parent2, best)\n","\n","    # Determine which parent is better\n","    better_parent, worse_parent = F2 < F1 ? (parent2, parent1) : (parent1, parent2)\n","\n","    # Initialize the first child (C1) as a copy of the worse parent\n","    C1 = deepcopy(worse_parent)\n","\n","    # Randomly determine the number of operations to apply\n","    #m = rand(1:min(50, maximum([length(chromosome) for chromosome in better_parent])))\n","    m = rand(1:5)\n","    C1 = get_operations(C1, better_parent, m)\n","    C2 = deepcopy(better_parent)\n","\n","    return C1, C2\n","end\n","\n"]},{"cell_type":"markdown","id":"ba560e44","metadata":{"id":"ba560e44"},"source":["**Mutate:**\n","\n","Mutation occurs in the population to induce some degree of random variety to break out of local minima. The mutate function finds the distance from the genome to the three genomes we are finding the median for, as well as the estimated median distance. A new genome is then generated on the sorting path between the input genome and whichever genome it's farthest away from the estimated distance of."]},{"cell_type":"code","execution_count":null,"id":"9c347057","metadata":{"id":"9c347057","outputId":"4734be8a-025b-427a-b30b-77a518dbf3f9"},"outputs":[{"data":{"text/plain":["mutate (generic function with 1 method)"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["function mutate(genomes, genome)\n","    # Calculate the DCJ distances between the leaves\n","    distance12 = dcj_distance(genomes[1], genomes[2])\n","    distance13 = dcj_distance(genomes[1], genomes[3])\n","    distance23 = dcj_distance(genomes[2], genomes[3])\n","\n","    # Calculate estimated distances from the median to the three given genomes\n","    distance1M = (distance12 + distance13 - distance23) ÷ 2\n","    distance2M = (distance12 + distance23 - distance13) ÷ 2\n","    distance3M = (distance13 + distance23 - distance12) ÷ 2\n","\n","    # Calculate actual distances from genome to the leaves\n","    dG1 = dcj_distance(genome, genomes[1])\n","    dG2 = dcj_distance(genome, genomes[2])\n","    dG3 = dcj_distance(genome, genomes[3])\n","\n","    # Calculate differences from estimated distances\n","    differences = [abs(dG1 - distance1M), abs(dG2 - distance2M), abs(dG3 - distance3M)]\n","    max_diff_index = argmax(differences)\n","\n","    # Select the target genome for mutation based on the greatest distance\n","    target_genome = genomes[max_diff_index]\n","\n","    # Choose a random number of operations for mutation\n","    m = rand(1:5)\n","\n","    # Apply m DCJ operations to mutate the genome\n","    mutated_genome = deepcopy(genome)\n","    mutated_genome = get_operations(mutated_genome, target_genome, m)\n","\n","    return mutated_genome\n","end\n"]},{"cell_type":"markdown","id":"ead24f09","metadata":{"id":"ead24f09"},"source":["**Read File:**\n","\n","Our read file function is designed to work with the input format of our tsv files. The file is read in and preprocessed to drop all irrelevant rows and columns, such as processing out mitochondrial data, which is highly preserved, and accession information, which is not usueful for our purposes."]},{"cell_type":"code","execution_count":null,"id":"884487d6","metadata":{"id":"884487d6","outputId":"e2019406-d156-43e8-be66-081b6a6fe9cb"},"outputs":[{"name":"stderr","output_type":"stream","text":["\u001b[32m\u001b[1m    Updating\u001b[22m\u001b[39m registry at `~/opt/anaconda3/envs/csc_530/share/julia/registries/General.toml`\n","\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n","\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/opt/anaconda3/envs/CSC_530/share/julia/environments/CSC_530/Project.toml`\n","\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/opt/anaconda3/envs/CSC_530/share/julia/environments/CSC_530/Manifest.toml`\n","\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n","\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/opt/anaconda3/envs/CSC_530/share/julia/environments/CSC_530/Project.toml`\n","\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/opt/anaconda3/envs/CSC_530/share/julia/environments/CSC_530/Manifest.toml`\n"]},{"data":{"text/plain":["read_file (generic function with 1 method)"]},"execution_count":14,"metadata":{},"output_type":"execute_result"}],"source":["import Pkg\n","Pkg.add(\"DataFrames\")\n","Pkg.add(\"CSV\")\n","using CSV\n","using DataFrames\n","\n","function read_file(filename)\n","    println(filename)\n","    data = CSV.File(filename, delim='\\t')\n","    println(\"read done\")\n","    dataframe = DataFrame(data)\n","    new_dataframe = copy(dataframe)\n","\n","    new_dataframe = filter(row -> occursin(\"protein-coding\", row[Symbol(\"Gene Type\")]), new_dataframe)\n","    new_dataframe = filter(row -> !occursin(\"uncharacterized gene\", row[Symbol(\"Name\")]), new_dataframe)\n","\n","    # Check if \"Chromosome\" is missing before calling occursin\n","    new_dataframe = filter(row -> !ismissing(row[Symbol(\"Chromosome\")]) &&\n","    !occursin(\"MT\", row[Symbol(\"Chromosome\")]), new_dataframe)\n","\n","    # Use select! to modify the DataFrame in place rather than creating multiple intermediate copies\n","    select!(new_dataframe, Not([Symbol(\"Name\"), Symbol(\"Protein length\"), Symbol(\"Transcripts accession\"),\n","                                Symbol(\"Protein accession\"), Symbol(\"Accession\"), Symbol(\"Gene ID\"),\n","                                Symbol(\"Gene Type\")]))\n","    return new_dataframe\n","end"]},{"cell_type":"markdown","id":"c7d1e5c9","metadata":{"id":"c7d1e5c9"},"source":["**Read Genomes:**\n","\n","This alternative file reading function is used to upload the sample data from the plos paper into our genome representation format for testing purposes."]},{"cell_type":"code","execution_count":null,"id":"56991324","metadata":{"scrolled":true,"id":"56991324","outputId":"3dd53b8d-710f-4373-acc0-84f1f40ab870"},"outputs":[{"data":{"text/plain":["read_genomes (generic function with 1 method)"]},"execution_count":15,"metadata":{},"output_type":"execute_result"}],"source":["function read_genomes(filename)\n","    genomes = []\n","    current_genome = []\n","    current_chromosome = [0]  # Start with a telomere\n","\n","    open(filename, \"r\") do file\n","        for line in eachline(file)\n","            line = strip(line)\n","            if startswith(line, \">\")  # New genome\n","                if !isempty(current_chromosome) > 1\n","                    push!(current_chromosome, 0)  # End with a telomere\n","                    push!(current_genome, current_chromosome)\n","                end\n","                if !isempty(current_genome)\n","                    push!(genomes, current_genome)\n","                end\n","                current_genome = []\n","                current_chromosome = [0]\n","            elseif line != \"\"\n","                genes = split(line)\n","                for gene in genes\n","                    if gene == \"\\$\"\n","                        push!(current_chromosome, 0)  # End with a telomere\n","                        if !isempty(current_chromosome)\n","                            push!(current_genome, current_chromosome)\n","                        end\n","                        current_chromosome = [0]  # Start new chromosome with a telomere\n","                    elseif gene != \"\"\n","                        push!(current_chromosome, parse(Int, gene))\n","                    end\n","                end\n","            end\n","        end\n","\n","        # Handle the last chromosome and genome\n","        if length(current_chromosome) > 1\n","            push!(current_chromosome, 0)  # End with a telomere\n","            push!(current_genome, current_chromosome)\n","        end\n","        if !isempty(current_genome)\n","            push!(genomes, current_genome)\n","        end\n","    end\n","\n","    return genomes\n","end\n"]},{"cell_type":"markdown","id":"5f2aa4f7","metadata":{"id":"5f2aa4f7"},"source":["**Make Dict:**\n","\n","This function is used to help create the genome representation of our input genome data to use in our genetic algorithm. The function finds all genes that are shared by all three genomes, and then creates a dictionary mapping each one to a number. This allows us to encode a consistent representation across all three genomes."]},{"cell_type":"code","execution_count":null,"id":"0652432c","metadata":{"id":"0652432c","outputId":"07d6deba-b189-4dd8-b5d2-817498949828"},"outputs":[{"data":{"text/plain":["make_dict (generic function with 1 method)"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["function make_dict(df1, df2, df3)\n","    # Use intersection of sets for efficient lookup and to ensure uniqueness\n","    symbols_set1 = Set(df1[!, Symbol(\"Symbol\")])\n","    symbols_set2 = Set(df2[!, Symbol(\"Symbol\")])\n","    symbols_set3 = Set(df3[!, Symbol(\"Symbol\")])\n","\n","    # Find common symbols across all three dataframes\n","    common_symbols = intersect(symbols_set1, symbols_set2, symbols_set3)\n","\n","    # Create a dictionary from the common symbols with incremental encoding\n","    dict = Dict{String, Int}([(symbol, index) for (index, symbol) in enumerate(common_symbols)])\n","\n","    return dict\n","end\n"]},{"cell_type":"markdown","id":"8cee18ee","metadata":{"id":"8cee18ee"},"source":["**Encoding:**\n","\n","This function encodes each of our genomes into our representation format using a pre-computed dictionary mapping genomes to a number. A genome is represented as an array of arrays, with each inner array representing a chromosome. Each value in the chromosome array represents a gene based off the mapping in the dictionary. The gene can be positive for a plus orientation, or negative for a negative orientation. Telomeres are represented using a 0. Chromosome arrays that do not contain any telomeres are circular chromosomes."]},{"cell_type":"code","execution_count":null,"id":"87a70693","metadata":{"id":"87a70693","outputId":"b29110a3-d688-4e97-d8d1-1faf416f6fea"},"outputs":[{"data":{"text/plain":["encoding (generic function with 1 method)"]},"execution_count":17,"metadata":{},"output_type":"execute_result"}],"source":["function encoding(df, dict)\n","    sequences = []  # Array to store chromosomes\n","    curr_sequence = []  # Current chromosome sequence\n","    currChro = \"\"  # Current chromosome identifier\n","\n","    for row in eachrow(df)\n","        symbol = row[Symbol(\"Symbol\")]\n","        chro = row[Symbol(\"Chromosome\")]\n","        sign = row[Symbol(\"Orientation\")]\n","\n","        # Check if a new chromosome is encountered\n","        if chro != currChro\n","            # Add the current chromosome if it contains more than just a telomere\n","            if length(curr_sequence) > 1\n","                push!(curr_sequence, 0)  # Add a telomere to end the chromosome\n","                push!(sequences, curr_sequence)\n","            end\n","            # Start a new chromosome sequence\n","            curr_sequence = [0]  # Start with a telomere\n","            currChro = chro\n","        end\n","\n","        # Add the gene encoding to the current chromosome\n","        if haskey(dict, symbol)\n","            gene_code = dict[symbol]›\n","            append!(curr_sequence, sign == \"plus\" ? gene_code : -gene_code)\n","        end\n","    end\n","\n","    # Add the last chromosome if it contains more than just a telomere\n","    if length(curr_sequence) > 1\n","        push!(curr_sequence, 0)  # End with a telomere\n","        push!(sequences, curr_sequence)\n","    end\n","\n","    return sequences\n","end\n","\n"]},{"cell_type":"markdown","id":"afc50ea7","metadata":{"id":"afc50ea7"},"source":["Read in big cat data file"]},{"cell_type":"code","execution_count":null,"id":"3e4485c0","metadata":{"id":"3e4485c0","outputId":"122b2645-30d8-4eb0-8899-ddd08793a852"},"outputs":[{"name":"stdout","output_type":"stream","text":["snowLeopard_raw.tsv\n","read done\n","lion_raw.tsv\n","read done\n","tiger_raw.tsv\n","read done\n"]},{"data":{"text/html":["<div><div style = \"float: left;\"><span>20219×5 DataFrame</span></div><div style = \"float: right;\"><span style = \"font-style: italic;\">20194 rows omitted</span></div><div style = \"clear: both;\"></div></div><div class = \"data-frame\" style = \"overflow-x: scroll;\"><table class = \"data-frame\" style = \"margin-bottom: 6px;\"><thead><tr class = \"header\"><th class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">Row</th><th style = \"text-align: left;\">Begin</th><th style = \"text-align: left;\">End</th><th style = \"text-align: left;\">Chromosome</th><th style = \"text-align: left;\">Orientation</th><th style = \"text-align: left;\">Symbol</th></tr><tr class = \"subheader headerLastRow\"><th class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\"></th><th title = \"Int64\" style = \"text-align: left;\">Int64</th><th title = \"Int64\" style = \"text-align: left;\">Int64</th><th title = \"Union{Missing, String3}\" style = \"text-align: left;\">String3?</th><th title = \"String7\" style = \"text-align: left;\">String7</th><th title = \"Union{Missing, String15}\" style = \"text-align: left;\">String15?</th></tr></thead><tbody><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">1</td><td style = \"text-align: right;\">7619</td><td style = \"text-align: right;\">21710</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC102958199</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">2</td><td style = \"text-align: right;\">77909</td><td style = \"text-align: right;\">81877</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">minus</td><td style = \"text-align: left;\">CCDC115</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">3</td><td style = \"text-align: right;\">81962</td><td style = \"text-align: right;\">86631</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">IMP4</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">4</td><td style = \"text-align: right;\">96735</td><td style = \"text-align: right;\">115102</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">PTPN18</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">5</td><td style = \"text-align: right;\">167025</td><td style = \"text-align: right;\">169922</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122230760</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">6</td><td style = \"text-align: right;\">170480</td><td style = \"text-align: right;\">188441</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122241520</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">7</td><td style = \"text-align: right;\">188614</td><td style = \"text-align: right;\">200720</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122241522</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">8</td><td style = \"text-align: right;\">201764</td><td style = \"text-align: right;\">224565</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC102952371</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">9</td><td style = \"text-align: right;\">305069</td><td style = \"text-align: right;\">333952</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">ATP12A</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">10</td><td style = \"text-align: right;\">368070</td><td style = \"text-align: right;\">480725</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">RNF17</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">11</td><td style = \"text-align: right;\">482391</td><td style = \"text-align: right;\">537460</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">minus</td><td style = \"text-align: left;\">CENPJ</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">12</td><td style = \"text-align: right;\">542108</td><td style = \"text-align: right;\">650756</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">PARP4</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">13</td><td style = \"text-align: right;\">655772</td><td style = \"text-align: right;\">707735</td><td style = \"text-align: left;\">A1</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">MPHOSPH8</td></tr><tr><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td><td style = \"text-align: right;\">&vellip;</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20208</td><td style = \"text-align: right;\">61951</td><td style = \"text-align: right;\">62547</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">minus</td><td style = \"text-align: left;\">LOC122236310</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20209</td><td style = \"text-align: right;\">177285</td><td style = \"text-align: right;\">185433</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">minus</td><td style = \"text-align: left;\">LOC122236311</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20210</td><td style = \"text-align: right;\">128557</td><td style = \"text-align: right;\">173509</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122236312</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20211</td><td style = \"text-align: right;\">82205</td><td style = \"text-align: right;\">83880</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122236314</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20212</td><td style = \"text-align: right;\">8773</td><td style = \"text-align: right;\">40299</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">minus</td><td style = \"text-align: left;\">LOC122236320</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20213</td><td style = \"text-align: right;\">40411</td><td style = \"text-align: right;\">83601</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122236316</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20214</td><td style = \"text-align: right;\">82487</td><td style = \"text-align: right;\">107753</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122236315</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20215</td><td style = \"text-align: right;\">160291</td><td style = \"text-align: right;\">161520</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122236318</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20216</td><td style = \"text-align: right;\">2650</td><td style = \"text-align: right;\">105695</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC102955594</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20217</td><td style = \"text-align: right;\">128623</td><td style = \"text-align: right;\">163648</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC107180711</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20218</td><td style = \"text-align: right;\">2026</td><td style = \"text-align: right;\">63101</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">minus</td><td style = \"text-align: left;\">LOC122236323</td></tr><tr><td class = \"rowNumber\" style = \"font-weight: bold; text-align: right;\">20219</td><td style = \"text-align: right;\">14142</td><td style = \"text-align: right;\">71585</td><td style = \"text-align: left;\">F3</td><td style = \"text-align: left;\">plus</td><td style = \"text-align: left;\">LOC122236330</td></tr></tbody></table></div>"],"text/latex":["\\begin{tabular}{r|ccccc}\n","\t& Begin & End & Chromosome & Orientation & Symbol\\\\\n","\t\\hline\n","\t& Int64 & Int64 & String3? & String7 & String15?\\\\\n","\t\\hline\n","\t1 & 7619 & 21710 & A1 & plus & LOC102958199 \\\\\n","\t2 & 77909 & 81877 & A1 & minus & CCDC115 \\\\\n","\t3 & 81962 & 86631 & A1 & plus & IMP4 \\\\\n","\t4 & 96735 & 115102 & A1 & plus & PTPN18 \\\\\n","\t5 & 167025 & 169922 & A1 & plus & LOC122230760 \\\\\n","\t6 & 170480 & 188441 & A1 & plus & LOC122241520 \\\\\n","\t7 & 188614 & 200720 & A1 & plus & LOC122241522 \\\\\n","\t8 & 201764 & 224565 & A1 & plus & LOC102952371 \\\\\n","\t9 & 305069 & 333952 & A1 & plus & ATP12A \\\\\n","\t10 & 368070 & 480725 & A1 & plus & RNF17 \\\\\n","\t11 & 482391 & 537460 & A1 & minus & CENPJ \\\\\n","\t12 & 542108 & 650756 & A1 & plus & PARP4 \\\\\n","\t13 & 655772 & 707735 & A1 & plus & MPHOSPH8 \\\\\n","\t14 & 741351 & 814430 & A1 & minus & PSPC1 \\\\\n","\t15 & 834714 & 868558 & A1 & minus & ZMYM5 \\\\\n","\t16 & 941754 & 1045558 & A1 & plus & ZMYM2 \\\\\n","\t17 & 1103792 & 1122253 & A1 & minus & GJA3 \\\\\n","\t18 & 1148487 & 1186272 & A1 & minus & GJB2 \\\\\n","\t19 & 1187264 & 1193807 & A1 & minus & GJB6 \\\\\n","\t20 & 1370374 & 1502847 & A1 & minus & CRYL1 \\\\\n","\t21 & 1529733 & 1661651 & A1 & plus & IFT88 \\\\\n","\t22 & 1670266 & 1689829 & A1 & plus & IL17D \\\\\n","\t23 & 1694690 & 1719209 & A1 & minus & EEF1AKMT1 \\\\\n","\t24 & 1728281 & 1842389 & A1 & minus & XPO4 \\\\\n","\t25 & 1897806 & 1974049 & A1 & minus & LATS2 \\\\\n","\t26 & 2065314 & 2071359 & A1 & plus & SAP18 \\\\\n","\t27 & 2076028 & 2096793 & A1 & minus & SKA3 \\\\\n","\t28 & 2096843 & 2097897 & A1 & plus & MRPL57 \\\\\n","\t29 & 2118065 & 2186219 & A1 & minus & ZDHHC20 \\\\\n","\t30 & 2215003 & 2358014 & A1 & minus & MICU2 \\\\\n","\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ \\\\\n","\\end{tabular}\n"],"text/plain":["\u001b[1m20219×5 DataFrame\u001b[0m\n","\u001b[1m   Row \u001b[0m│\u001b[1m Begin  \u001b[0m\u001b[1m End    \u001b[0m\u001b[1m Chromosome \u001b[0m\u001b[1m Orientation \u001b[0m\u001b[1m Symbol       \u001b[0m\n","       │\u001b[90m Int64  \u001b[0m\u001b[90m Int64  \u001b[0m\u001b[90m String3?   \u001b[0m\u001b[90m String7     \u001b[0m\u001b[90m String15?    \u001b[0m\n","───────┼───────────────────────────────────────────────────────\n","     1 │   7619   21710  A1          plus         LOC102958199\n","     2 │  77909   81877  A1          minus        CCDC115\n","     3 │  81962   86631  A1          plus         IMP4\n","     4 │  96735  115102  A1          plus         PTPN18\n","     5 │ 167025  169922  A1          plus         LOC122230760\n","     6 │ 170480  188441  A1          plus         LOC122241520\n","     7 │ 188614  200720  A1          plus         LOC122241522\n","     8 │ 201764  224565  A1          plus         LOC102952371\n","     9 │ 305069  333952  A1          plus         ATP12A\n","    10 │ 368070  480725  A1          plus         RNF17\n","    11 │ 482391  537460  A1          minus        CENPJ\n","   ⋮   │   ⋮       ⋮         ⋮            ⋮            ⋮\n"," 20210 │ 128557  173509  F3          plus         LOC122236312\n"," 20211 │  82205   83880  F3          plus         LOC122236314\n"," 20212 │   8773   40299  F3          minus        LOC122236320\n"," 20213 │  40411   83601  F3          plus         LOC122236316\n"," 20214 │  82487  107753  F3          plus         LOC122236315\n"," 20215 │ 160291  161520  F3          plus         LOC122236318\n"," 20216 │   2650  105695  F3          plus         LOC102955594\n"," 20217 │ 128623  163648  F3          plus         LOC107180711\n"," 20218 │   2026   63101  F3          minus        LOC122236323\n"," 20219 │  14142   71585  F3          plus         LOC122236330\n","\u001b[36m                                             20198 rows omitted\u001b[0m"]},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["snow_leopard_file = \"snowLeopard_raw.tsv\"\n","tiger_file = \"tiger_raw.tsv\"\n","lion_file = \"lion_raw.tsv\"\n","\n","sl_df = read_file(snow_leopard_file)\n","lion_df = read_file(lion_file)\n","tiger_df = read_file(tiger_file)"]},{"cell_type":"code","execution_count":null,"id":"30218bfd","metadata":{"id":"30218bfd","outputId":"64e910cd-b0d4-4b55-fc4a-d240bbee1370"},"outputs":[{"data":{"text/plain":["Dict{String, Int64} with 14740 entries:\n","  \"VPS41\"    => 14740\n","  \"SHTN1\"    => 7379\n","  \"TMOD4\"    => 2\n","  \"ETV4\"     => 3\n","  \"IQCF6\"    => 7380\n","  \"HOXA9\"    => 7381\n","  \"NKAIN3\"   => 7382\n","  \"SMTNL2\"   => 7383\n","  \"FAM124A\"  => 7384\n","  \"IRS2\"     => 7385\n","  \"DNASE2\"   => 4\n","  \"COLEC11\"  => 7386\n","  \"TMSB4X\"   => 7387\n","  \"GRM7\"     => 7388\n","  \"EIF4EBP3\" => 7389\n","  \"KAT2A\"    => 5\n","  \"MORC2\"    => 7390\n","  \"IL17F\"    => 7391\n","  \"ZSWIM7\"   => 7392\n","  \"TMEM208\"  => 7393\n","  \"SREBF2\"   => 6\n","  \"CDC40\"    => 7\n","  \"IL12RB2\"  => 7394\n","  \"TEX49\"    => 8\n","  \"FAXC\"     => 9\n","  ⋮          => ⋮"]},"execution_count":19,"metadata":{},"output_type":"execute_result"}],"source":["dict = make_dict(sl_df, tiger_df, lion_df)"]},{"cell_type":"code","execution_count":null,"id":"86591f1d","metadata":{"id":"86591f1d","outputId":"98eeae3e-9012-4c2c-cc15-b33bba6573d8"},"outputs":[{"data":{"text/plain":["20-element Vector{Any}:\n"," [0, -8420, 2258, 5473, 11465, 2465, -11749, 6424, 1202, -9120  …  6479, -11336, 6512, -7052, -11210, -1291, 8629, -12543, -8934, 0]\n"," [0, -13849, -12534, -9535, -6810, 1569, 1655, 6201, 2649, 12989  …  3338, -6264, 2067, 5282, -7403, -370, -12005, 14087, -13499, 0]\n"," [0, -6604, -3101, 7031, -5211, 834, 8198, -8572, 6126, -1634  …  6458, 2371, -3657, -10385, 13063, 11852, -1537, 419, 9671, 0]\n"," [0, -1266, 10783, -7664, 3132, 2282, 4651, 1493, 5425, -4239  …  -13307, 14014, 3031, -11501, 1313, -5091, 4648, -837, -2332, 0]\n"," [0, -8359, 6945, -1963, -619, 9015, -9569, 3682, 13761, 8997  …  -585, 1466, -2934, -13074, 4697, -5125, 14676, -12636, -8416, 0]\n"," [0, 13755, -13868, -10079, -3196, -831, -10405, -7505, 14466, -7588  …  -14226, -1365, -991, 7952, 9446, 1543, 235, 1514, 13271, 0]\n"," [0, 13438, -6248, -12505, 11161, -4675, 14582, -11917, -4928, -12299  …  7871, -11360, -8005, -7949, 8182, -838, 13243, 9436, -7533, 0]\n"," [0, 9520, -2106, 2455, 4268, -7167, -3934, 3765, 5659, -11474  …  -2169, 4717, -4394, 14508, 9722, 5693, 4816, -12857, 8338, 0]\n"," [0, -9252, 11052, -8672, -2087, 11542, -5823, 12107, 2559, 4226  …  -13298, -3874, 12828, -2464, 13091, 7092, 7269, -7544, 3573, 0]\n"," [0, 3036, -2523, -5302, 8354, 11678, 580, 437, -5419, -4885  …  7754, -3843, -2750, -495, 10201, -12261, 7536, -10584, -4659, 0]\n"," [0, 6675, 11855, -5051, -2208, 5347, -5416, -7305, 10521, 4409  …  -13410, 1729, -3881, -487, -9743, 12790, 5526, -5567, -3001, 0]\n"," [0, -14121, -8224, -1117, 8242, 3008, 2711, -13843, -13606, 8376  …  7206, 7255, 11939, -6554, 11793, -8007, 9876, 11761, -14478, 0]\n"," [0, 6309, -6007, -3156, -11785, -1671, -3017, -2175, -9269, -8615  …  -8879, -6566, -4919, 9793, -13552, -12379, 12241, 11006, 12886, 0]\n"," [0, 1567, 12671, -1746, -8585, 7383, 9539, -8760, 9771, 1132  …  6389, 4005, -950, -518, 6329, 3276, -6756, -10845, 3971, 0]\n"," [0, 14216, 1698, 6361, -9441, 11416, 2295, -3188, -7481, -3138  …  -6576, 7218, -1505, 9767, 10519, 3814, 13994, -8069, -13602, 0]\n"," [0, 4663, 6299, -8031, -5179, 13471, 13166, 12452, -11384, -8312  …  -7811, 13567, -1516, 1022, -11314, 8846, -6628, 10902, 1770, 0]\n"," [0, 5195, -8699, -5097, 12155, 8520, -11325, 8200, -179, -5279  …  -13263, 1348, -12320, -8566, 10911, 9607, -1504, -2920, 4667, 0]\n"," [0, 3976, -1637, -11532, 1502, 13554, -7057, 1354, 11908, -5667  …  13611, 13085, 13150, 385, 2748, 8778, 8569, 4188, -9739, 0]\n"," [0, -14358, -1486, 14242, 11058, -578, -6240, 1823, 9220, -9740  …  -2504, -6491, 6236, -9480, 9315, 3683, 6733, 10409, -7387, 0]\n"," [0, -5273, 0]"]},"execution_count":20,"metadata":{},"output_type":"execute_result"}],"source":["encoded_sl = encoding(sl_df, dict)\n","encoded_tiger = encoding(tiger_df, dict)\n","encoded_lion = encoding(lion_df, dict)"]},{"cell_type":"markdown","id":"8432d6d1","metadata":{"id":"8432d6d1"},"source":["This block is used for intial population generation as well as defining parameters. The sample variable contains the genomes from the reference paper, which are stored in the data.txt file. The commented out line defines genomes using the big cat genomes instead. Keep in mind that using the big cat data is significantly slower due to the size of the genomes."]},{"cell_type":"code","execution_count":null,"id":"c7efa380","metadata":{"id":"c7efa380","outputId":"ad8c5b1d-4b9a-4fdc-a00f-4a4dd8b7d09a"},"outputs":[{"name":"stdout","output_type":"stream","text":["1\n","2\n","3\n","4\n","5\n","6\n","7\n","8\n","9\n","10\n","11\n","12\n","13\n","14\n","15\n","16\n","17\n","18\n","19\n","20\n","21\n","22\n","23\n","24\n","25\n","26\n","27\n","28\n","29\n","30\n","31\n","32\n","33\n","34\n","35\n","36\n","37\n","38\n","39\n","40\n","41\n","42\n","43\n","44\n","45\n","46\n","47\n","48\n","49\n","50\n","51\n","52\n","53\n","54\n","55\n","56\n","57\n","58\n","59\n","60\n","61\n","62\n","63\n","64\n","65\n","66\n","67\n","68\n","69\n","70\n","71\n","72\n","73\n","74\n","75\n","76\n","77\n","78\n","79\n","80\n","81\n","82\n","83\n","84\n","85\n","86\n","87\n","88\n","89\n","90\n","91\n","92\n","93\n","94\n","95\n","96\n","97\n","98\n","99\n","100\n","101\n","102\n","103\n","104\n","105\n","106\n","107\n","108\n","109\n","110\n","111\n","112\n","113\n","114\n","115\n","116\n","117\n","118\n","119\n","120\n","121\n","122\n","123\n","124\n","125\n","126\n","127\n","128\n","129\n","130\n","131\n","132\n","133\n","134\n","135\n","136\n","137\n","138\n","139\n","140\n","141\n","142\n","143\n","144\n","145\n","146\n","147\n","148\n","149\n","150\n","151\n","152\n","153\n","154\n","155\n","156\n","157\n","158\n","159\n","160\n","161\n","162\n","163\n","164\n","165\n","166\n","167\n","168\n","169\n","170\n","171\n","172\n","173\n","174\n","175\n","176\n","177\n","178\n","179\n","180\n","181\n","182\n","183\n","184\n","185\n","186\n","187\n","188\n","189\n","190\n","191\n","192\n","193\n","194\n","195\n","196\n","197\n","198\n","199\n","200\n","201\n","202\n","203\n","204\n","205\n","206\n","207\n","208\n","209\n","210\n","211\n","212\n","213\n","214\n","215\n","216\n","217\n","218\n","219\n","220\n","221\n","222\n","223\n","224\n","225\n","226\n","227\n","228\n","229\n","230\n","231\n","232\n","233\n","234\n","235\n","236\n","237\n","238\n","239\n","240\n","241\n","242\n","243\n","244\n","245\n","246\n","247\n","248\n","249\n","250\n","251\n","252\n","253\n","254\n","255\n","256\n","257\n","258\n","259\n","260\n","261\n","262\n","263\n","264\n","265\n","266\n","267\n","268\n","269\n","270\n","271\n","272\n","273\n","274\n","275\n","276\n","277\n","278\n","279\n","280\n","281\n","282\n","283\n","284\n","285\n","286\n","287\n","288\n","289\n","290\n","291\n","292\n","293\n","294\n","295\n","296\n","297\n","298\n","299\n","300\n","301\n","302\n","303\n","304\n","305\n","306\n","307\n","308\n","309\n","310\n","311\n","312\n","313\n","314\n","315\n","316\n","317\n","318\n","319\n","320\n","321\n","322\n","323\n","324\n","325\n","326\n","327\n","328\n","329\n","330\n","331\n","332\n","333\n","334\n","335\n","336\n","337\n","338\n","339\n","340\n","341\n","342\n","343\n","344\n","345\n","346\n","347\n","348\n","349\n","350\n","351\n","352\n","353\n","354\n","355\n","356\n","357\n","358\n","359\n","360\n","361\n","362\n","363\n","364\n","365\n","366\n","367\n","368\n","369\n","370\n","371\n","372\n","373\n","374\n","375\n","376\n","377\n","378\n","379\n","380\n","381\n","382\n","383\n","384\n","385\n","386\n","387\n","388\n","389\n","390\n","391\n","392\n","393\n","394\n","395\n","396\n","397\n","398\n","399\n","400\n","401\n","402\n","403\n","404\n","405\n","406\n","407\n","408\n","409\n","410\n","411\n","412\n","413\n","414\n","415\n","416\n","417\n","418\n","419\n","420\n","421\n","422\n","423\n","424\n","425\n","426\n","427\n","428\n","429\n","430\n","431\n","432\n","433\n","434\n","435\n","436\n","437\n","438\n","439\n","440\n","441\n","442\n","443\n","444\n","445\n","446\n","447\n","448\n","449\n","450\n","451\n","452\n","453\n","454\n","455\n","456\n","457\n","458\n","459\n","460\n","461\n","462\n","463\n","464\n","465\n","466\n","467\n","468\n","469\n","470\n","471\n","472\n","473\n","474\n","475\n","476\n","477\n","478\n","479\n","480\n","481\n","482\n","483\n","484\n","485\n","486\n","487\n","488\n","489\n","490\n","491\n","492\n","493\n","494\n","495\n","496\n","497\n","498\n","499\n","500\n","501\n","502\n","503\n","504\n","505\n","506\n","507\n","508\n","509\n","510\n","511\n","512\n","513\n","514\n","515\n","516\n","517\n","518\n","519\n","520\n","521\n","522\n","523\n","524\n","525\n","526\n","527\n","528\n","529\n","530\n","531\n","532\n","533\n","534\n","535\n","536\n","537\n","538\n","539\n","540\n","541\n","542\n","543\n","544\n","545\n","546\n","547\n","548\n","549\n","550\n","551\n","552\n","553\n","554\n","555\n","556\n","557\n","558\n","559\n","560\n","561\n","562\n","563\n","564\n","565\n","566\n","567\n","568\n","569\n","570\n","571\n","572\n","573\n","574\n","575\n","576\n","577\n","578\n","579\n","580\n","581\n","582\n","583\n","584\n","585\n","586\n","587\n","588\n","589\n","590\n","591\n","592\n","593\n","594\n","595\n","596\n","597\n","598\n","599\n","600\n","601\n","602\n","603\n","604\n","605\n","606\n","607\n","608\n","609\n","610\n","611\n","612\n","613\n","614\n","615\n","616\n","617\n","618\n","619\n","620\n","621\n","622\n","623\n","624\n","625\n","626\n","627\n","628\n","629\n","630\n","631\n","632\n","633\n","634\n","635\n","636\n","637\n","638\n","639\n","640\n","641\n","642\n","643\n","644\n","645\n","646\n","647\n","648\n","649\n","650\n","651\n","652\n","653\n","654\n","655\n","656\n","657\n","658\n","659\n","660\n","661\n","662\n","663\n","664\n","665\n","666\n","667\n","668\n","669\n","670\n","671\n","672\n","673\n","674\n","675\n","676\n","677\n","678\n","679\n","680\n","681\n","682\n","683\n","684\n","685\n","686\n","687\n","688\n","689\n","690\n","691\n","692\n","693\n","694\n","695\n","696\n","697\n","698\n","699\n","700\n","701\n","702\n","703\n","704\n","705\n","706\n","707\n","708\n","709\n","710\n","711\n","712\n","713\n","714\n","715\n","716\n","717\n","718\n","719\n","720\n","721\n","722\n","723\n","724\n","725\n","726\n","727\n","728\n","729\n","730\n","731\n","732\n","733\n","734\n","735\n","736\n","737\n","738\n","739\n","740\n","741\n","742\n","743\n","744\n","745\n","746\n","747\n","748\n","749\n","750\n","751\n","752\n","753\n","754\n","755\n","756\n","757\n","758\n","759\n","760\n","761\n","762\n","763\n","764\n","765\n","766\n","767\n","768\n","769\n","770\n","771\n","772\n","773\n","774\n","775\n","776\n","777\n","778\n","779\n","780\n","781\n","782\n","783\n","784\n","785\n","786\n","787\n","788\n","789\n","790\n","791\n","792\n","793\n","794\n","795\n","796\n","797\n","798\n","799\n","800\n","801\n","802\n","803\n","804\n","805\n","806\n","807\n","808\n","809\n","810\n","811\n","812\n","813\n","814\n","815\n","816\n","817\n","818\n","819\n","820\n","821\n","822\n","823\n","824\n","825\n","826\n","827\n","828\n","829\n","830\n","831\n","832\n","833\n","834\n","835\n","836\n","837\n","838\n","839\n","840\n","841\n","842\n","843\n","844\n","845\n","846\n","847\n","848\n","849\n","850\n","851\n","852\n","853\n","854\n","855\n","856\n","857\n","858\n","859\n","860\n","861\n","862\n","863\n","864\n","865\n","866\n","867\n","868\n","869\n","870\n","871\n","872\n","873\n","874\n","875\n","876\n","877\n","878\n","879\n","880\n","881\n","882\n","883\n","884\n","885\n","886\n","887\n","888\n","889\n","890\n","891\n","892\n","893\n","894\n","895\n","896\n","897\n","898\n","899\n","900\n","901\n","902\n","903\n","904\n","905\n","906\n","907\n","908\n","909\n","910\n","911\n","912\n","913\n","914\n","915\n","916\n","917\n","918\n","919\n","920\n","921\n","922\n","923\n","924\n","925\n","926\n","927\n","928\n","929\n","930\n","931\n","932\n","933\n","934\n","935\n","936\n","937\n","938\n","939\n","940\n","941\n","942\n","943\n","944\n","945\n","946\n","947\n","948\n","949\n","950\n","951\n","952\n","953\n","954\n","955\n","956\n","957\n","958\n","959\n","960\n","961\n","962\n","963\n","964\n","965\n","966\n","967\n","968\n","969\n","970\n","971\n","972\n","973\n","974\n","975\n","976\n","977\n","978\n","979\n","980\n","981\n","982\n","983\n","984\n","985\n","986\n","987\n","988\n","989\n","990\n","991\n","992\n","993\n","994\n","995\n","996\n","997\n","998\n","999\n","1000\n","1001\n","1002\n","1003\n","1004\n","1005\n","1006\n","1007\n","1008\n","1009\n","1010\n","1011\n","1012\n","1013\n","1014\n","1015\n","1016\n","1017\n","1018\n","1019\n","1020\n","1021\n","1022\n","1023\n","1024\n","1025\n","1026\n","1027\n","1028\n","1029\n","1030\n","1031\n","1032\n","1033\n","1034\n","1035\n","1036\n","1037\n","1038\n","1039\n","1040\n","1041\n","1042\n","1043\n","1044\n","1045\n","1046\n","1047\n","1048\n","1049\n","1050\n","1051\n","1052\n","1053\n","1054\n","1055\n","1056\n","1057\n","1058\n","1059\n","1060\n","1061\n","1062\n","1063\n","1064\n","1065\n","1066\n","1067\n","1068\n","1069\n","1070\n","1071\n","1072\n","1073\n","1074\n","1075\n","1076\n","1077\n","1078\n","1079\n","1080\n","1081\n","1082\n","1083\n","1084\n","1085\n","1086\n","1087\n","1088\n","1089\n","1090\n","1091\n","1092\n","1093\n","1094\n","1095\n","1096\n","1097\n","1098\n","1099\n","1100\n","1101\n","1102\n","1103\n","1104\n","1105\n","1106\n","1107\n","1108\n","1109\n","1110\n","1111\n","1112\n","1113\n","1114\n","1115\n","1116\n","1117\n","1118\n","1119\n","1120\n","1121\n","1122\n","1123\n","1124\n","1125\n","1126\n","1127\n","1128\n","1129\n","1130\n","1131\n","1132\n","1133\n","1134\n","1135\n","1136\n","1137\n","1138\n","1139\n","1140\n","1141\n","1142\n","1143\n","1144\n","1145\n","1146\n","1147\n","1148\n","1149\n","1150\n","1151\n","1152\n","1153\n","1154\n","1155\n","1156\n","1157\n","1158\n","1159\n","1160\n","1161\n","1162\n","1163\n","1164\n","1165\n","1166\n","1167\n","1168\n","1169\n","1170\n","1171\n","1172\n","1173\n","1174\n","1175\n","1176\n","1177\n","1178\n","1179\n","1180\n","1181\n","1182\n","1183\n","1184\n","1185\n","1186\n","1187\n","1188\n","1189\n","1190\n","1191\n","1192\n","1193\n","1194\n","1195\n","1196\n","1197\n","1198\n","1199\n","1200\n","1201\n","1202\n","1203\n","1204\n","1205\n","1206\n","1207\n","1208\n","1209\n","1210\n","1211\n","1212\n","1213\n","1214\n","1215\n","1216\n","1217\n","1218\n","1219\n","1220\n","1221\n","1222\n","1223\n","1224\n","1225\n","1226\n","1227\n","1228\n","1229\n","1230\n","1231\n","1232\n","1233\n","1234\n","1235\n","1236\n","1237\n","1238\n","1239\n","1240\n","1241\n","1242\n","1243\n","1244\n","1245\n","1246\n","1247\n","1248\n","1249\n","1250\n","1251\n","1252\n","1253\n","1254\n","1255\n","1256\n","1257\n","1258\n","1259\n","1260\n","1261\n","1262\n","1263\n","1264\n","1265\n","1266\n","1267\n","1268\n","1269\n","1270\n","1271\n","1272\n","1273\n","1274\n","1275\n","1276\n","1277\n","1278\n","1279\n","1280\n","1281\n","1282\n","1283\n","1284\n","1285\n","1286\n","1287\n","1288\n","1289\n","1290\n","1291\n","1292\n","1293\n","1294\n","1295\n","1296\n","1297\n","1298\n","1299\n","1300\n","1301\n","1302\n","1303\n","1304\n","1305\n","1306\n","1307\n","1308\n","1309\n","1310\n","1311\n","1312\n","1313\n","1314\n","1315\n","1316\n","1317\n","1318\n","1319\n","1320\n","1321\n","1322\n","1323\n","1324\n","1325\n","1326\n","1327\n","1328\n","1329\n","1330\n","1331\n","1332\n","1333\n","1334\n","1335\n","1336\n","1337\n","1338\n","1339\n","1340\n","1341\n","1342\n","1343\n","1344\n","1345\n","1346\n","1347\n","1348\n","1349\n","1350\n","1351\n","1352\n","1353\n","1354\n","1355\n","1356\n","1357\n","1358\n","1359\n","1360\n","1361\n","1362\n","1363\n","1364\n","1365\n","1366\n","1367\n","1368\n","1369\n","1370\n","1371\n","1372\n","1373\n","1374\n","1375\n","1376\n","1377\n","1378\n","1379\n","1380\n","1381\n","1382\n","1383\n","1384\n","1385\n","1386\n","1387\n","1388\n","1389\n","1390\n","1391\n","1392\n","1393\n","1394\n","1395\n","1396\n","1397\n","1398\n","1399\n","1400\n","1401\n","1402\n","1403\n","1404\n","1405\n","1406\n","1407\n","1408\n","1409\n","1410\n","1411\n","1412\n","1413\n","1414\n","1415\n","1416\n","1417\n","1418\n","1419\n","1420\n","1421\n","1422\n","1423\n","1424\n","1425\n","1426\n","1427\n","1428\n","1429\n","1430\n","1431\n","1432\n","1433\n","1434\n","1435\n","1436\n","1437\n","1438\n","1439\n","1440\n","1441\n","1442\n","1443\n","1444\n","1445\n","1446\n","1447\n","1448\n","1449\n","1450\n","1451\n","1452\n","1453\n","1454\n","1455\n","1456\n","1457\n","1458\n","1459\n","1460\n","1461\n","1462\n","1463\n","1464\n","1465\n","1466\n","1467\n","1468\n","1469\n","1470\n","1471\n","1472\n","1473\n","1474\n","1475\n","1476\n","1477\n","1478\n","1479\n","1480\n","1481\n","1482\n","1483\n","1484\n","1485\n","1486\n","1487\n","1488\n","1489\n","1490\n","1491\n","1492\n","1493\n","1494\n","1495\n","1496\n","1497\n","1498\n","1499\n","1500\n","1501\n","1502\n","1503\n","1504\n","1505\n","1506\n","1507\n","1508\n","1509\n","1510\n","1511\n","1512\n","1513\n","1514\n","1515\n","1516\n","1517\n","1518\n","1519\n","1520\n","1521\n","1522\n","1523\n","1524\n","1525\n","1526\n","1527\n","1528\n","1529\n","1530\n","1531\n","1532\n","1533\n","1534\n","1535\n","1536\n","1537\n","1538\n","1539\n","1540\n","1541\n","1542\n","1543\n","1544\n","1545\n","1546\n","1547\n","1548\n","1549\n","1550\n","1551\n","1552\n","1553\n","1554\n","1555\n","1556\n","1557\n","1558\n","1559\n","1560\n","1561\n","1562\n","1563\n","1564\n","1565\n","1566\n","1567\n","1568\n","1569\n","1570\n","1571\n","1572\n","1573\n","1574\n","1575\n","1576\n","1577\n","1578\n","1579\n","1580\n","1581\n","1582\n","1583\n","1584\n","1585\n","1586\n","1587\n","1588\n","1589\n","1590\n","1591\n","1592\n","1593\n","1594\n","1595\n","1596\n","1597\n","1598\n","1599\n","1600\n","done\n"]}],"source":["#genomes = [encoded_sl, encoded_tiger, encoded_lion]\n","sample = read_genomes(\"data.txt\")\n","genomes = [sample[1],sample[2],sample[3]]\n","per_step = 100 #Can be adjusted to control the variance in the initial population\n","pop = initial_population(genomes, per_step)\n","\n","max_generations = 25 #Can be adjusted for testing purposes. Reference paper uses 500\n","best_genome = nothing\n","top_score = Inf\n","\n","perfect_median_score = best_score(genomes[1],genomes[2],genomes[3])\n","println(\"done\")"]},{"cell_type":"markdown","id":"46b45f16","metadata":{"id":"46b45f16"},"source":["This is the main loop of the program. This loop performs the genetic algorithm by creating a new population each iteration by performing crossover and mutations on each genome in the population. When the loop has either completed the maximum number of iterations or converged on the median genome, the highest performing genome is reported, and a plot showing the top score in the population vs iteration number is shown."]},{"cell_type":"code","execution_count":null,"id":"c94fcee2","metadata":{"id":"c94fcee2","outputId":"fb727c43-0743-44ab-d2d3-6605e7a9f71f"},"outputs":[{"name":"stderr","output_type":"stream","text":["\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n","\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/opt/anaconda3/envs/CSC_530/share/julia/environments/CSC_530/Project.toml`\n","\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/opt/anaconda3/envs/CSC_530/share/julia/environments/CSC_530/Manifest.toml`\n"]},{"name":"stdout","output_type":"stream","text":["25\n","Getting scores: \n","---------------------"]},{"ename":"LoadError","evalue":"InterruptException:","output_type":"error","traceback":["InterruptException:","","Stacktrace:"," [1] adjacency_graph(genome1::Vector{Any}, genome2::Vector{Any})","   @ Main ./In[2]:50"," [2] dcj_distance(genome1::Vector{Any}, genome2::Vector{Any})","   @ Main ./In[5]:4"," [3] median_score(genome1::Vector{Any}, genome2::Vector{Any}, genome3::Vector{Any}, m::Vector{Any})","   @ Main ./In[9]:4"," [4] fitness(genome1::Vector{Any}, genome2::Vector{Any}, genome3::Vector{Any}, m::Vector{Any}, best::Float64)","   @ Main ./In[10]:4"," [5] get_scores(genomes::Vector{Vector{Any}}, pop::Vector{Any}, best::Float64)","   @ Main ./In[11]:6"]}],"source":["Pkg.add(\"BenchmarkTools\")\n","using BenchmarkTools\n","using Plots\n","\n","top_scores = [] #Used for the plot\n","\n","println(max_generations)\n","#benchmark_result = @benchmark\n","for i in 1:max_generations\n","        println(\"Getting scores: \")\n","        flush(stdout)\n","        fitness_dict, sorted_genomes = get_scores(genomes, pop, perfect_median_score)\n","\n","        # Check for the best score in this generation\n","        current_top_score = fitness_dict[sorted_genomes[1]]\n","        if current_top_score < top_score\n","            top_score = current_top_score\n","            best_genome = sorted_genomes[1]\n","        end\n","\n","        push!(top_scores, top_score)\n","\n","        # Early termination if perfect median score is met\n","        if top_score == perfect_median_score  # Define 'perfect_median_score' as per your criteria\n","            println(\"Perfect median score met at generation $i\")\n","            break\n","        end\n","\n","        new_pop = []\n","\n","        #Select the top 10% of individuals\n","        top_10_percent = ceil(Int, length(sorted_genomes) * 0.1)\n","        new_pop = [deepcopy(genome) for genome in sorted_genomes[1:top_10_percent]]\n","\n","        # Selection for crossover from the remaining 90%\n","        for j in 1:((length(sorted_genomes) - top_10_percent))\n","            parent1 = sorted_genomes[rand(top_10_percent+1:end)]\n","            parent2 = sorted_genomes[rand(top_10_percent+1:end)]\n","            println(\"crossover\")\n","            C1, C2 = crossover(parent1, parent2, genomes..., perfect_median_score)\n","\n","            # Mutation\n","            println(\"mutate\")\n","            C1s = mutate(genomes, C1)\n","            C2s = mutate(genomes, C2)\n","\n","            # Evaluate fitness for each child\n","            fitness_scores = [\n","                (C1, fitness(genomes[1], genomes[2], genomes[3], C1, perfect_median_score)),\n","                (C2, fitness(genomes[1], genomes[2], genomes[3], C2, perfect_median_score)),\n","                (C1s, fitness(genomes[1], genomes[2], genomes[3], C1s, perfect_median_score)),\n","                (C2s, fitness(genomes[1], genomes[2], genomes[3], C2s, perfect_median_score))\n","            ]\n","\n","\n","            # Sort by fitness and select the top two\n","            best_two = sort(fitness_scores, by = x -> x[2], rev = true)[1:2]\n","\n","            # Add the best two offspring to new population\n","            push!(new_pop, best_two[1][1])\n","            push!(new_pop, best_two[2][1])\n","\n","            println(j)\n","        end\n","\n","        # Replace the old population with the new one\n","        pop = new_pop\n","        println(i)\n","        flush(stdout)\n","end #samples=1 evals=1\n","\n","#println(benchmark_result)\n","\n","\n","println(\"Best genome found: $best_genome with score: $top_score\")\n","plot(1:max_generations, top_scores, title=\"Fitness Score vs Iteration\", xlabel=\"Iteration\", ylabel=\"Fitness Score\", legend=false)"]},{"cell_type":"markdown","id":"2f8131e4","metadata":{"id":"2f8131e4"},"source":["This block defines a dictionary mapping and a function for the teaching staff to be able to convert the standardized genome format into our genome format for testing purposes."]},{"cell_type":"code","execution_count":null,"id":"404276a2","metadata":{"id":"404276a2","outputId":"d682e677-ff0f-4779-d3ed-9c47f31fc037"},"outputs":[{"name":"stdout","output_type":"stream","text":["Any[[0, 1, 2, 3, 0]]\n","Any[[1, 2, -3], [0, 6, 7, -8, 0], [25, -26]]\n"]}],"source":["# Global dictionary for character to gene mapping\n","global_map = Dict(\n","    'a' => 1, 'b' => 2, 'c' => 3, 'd' => 4, 'e' => 5, 'f' => 6, 'g' => 7, 'h' => 8, 'i' => 9, 'j' => 10,\n","    'k' => 11, 'l' => 12, 'm' => 13, 'n' => 14, 'o' => 15, 'p' => 16, 'q' => 17, 'r' => 18, 's' => 19, 't' => 20,\n","    'u' => 21, 'v' => 22, 'w' => 23, 'x' => 24, 'y' => 25, 'z' => 26,\n","    'A' => -1, 'B' => -2, 'C' => -3, 'D' => -4, 'E' => -5, 'F' => -6, 'G' => -7, 'H' => -8, 'I' => -9, 'J' => -10,\n","    'K' => -11, 'L' => -12, 'M' => -13, 'N' => -14, 'O' => -15, 'P' => -16, 'Q' => -17, 'R' => -18, 'S' => -19, 'T' => -20,\n","    'U' => -21, 'V' => -22, 'W' => -23, 'X' => -24, 'Y' => -25, 'Z' => -26\n",")\n","function convert_to_genome(input)\n","    # Function to convert a character to its gene number\n","    function char_to_int(c)\n","        if c == '.'\n","            return 0\n","        else\n","            return global_map[c]\n","        end\n","    end\n","\n","    # Convert input to genome format\n","    genome = []\n","    for chromosome in input\n","        push!(genome, [char_to_int(c) for c in chromosome])\n","    end\n","\n","    return genome\n","end\n","\n","# Example usage\n","input1 = [\".abc.\"]\n","input2 = [\"abC\", \".fgH.\", \"yZ\"]\n","genome1 = convert_to_genome(input1)\n","genome2 = convert_to_genome(input2)\n","\n","println(genome1)\n","println(genome2)\n"]},{"cell_type":"markdown","id":"ad30657a","metadata":{"id":"ad30657a"},"source":["Verification of our DCJ distance using the example inputs given on Piaza"]},{"cell_type":"code","execution_count":null,"id":"45e4f2c7","metadata":{"id":"45e4f2c7","outputId":"596eeb57-c746-4b84-d441-7dfac8f33d1e"},"outputs":[{"name":"stdout","output_type":"stream","text":["1\n","1\n","3\n"]}],"source":["A = convert_to_genome([\".ab.\"])\n","B = convert_to_genome([\"ab\"])\n","println(dcj_distance(A, B))\n","\n","A = convert_to_genome([\"ab\", \".cd.\"])\n","B = convert_to_genome([\".abcd.\"])\n","println(dcj_distance(A, B))\n","\n","A = convert_to_genome([\".abC.\", \"de\"])\n","B = convert_to_genome([\".abdCe.\"])\n","println(dcj_distance(A, B))"]}],"metadata":{"kernelspec":{"display_name":"Julia 1.8.5","language":"julia","name":"julia-1.8"},"language_info":{"file_extension":".jl","mimetype":"application/julia","name":"julia","version":"1.8.5"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}